lib/ddtrace/profiling/trace_identifiers/ddtrace.rb

```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+
+module Datadog
+  module Profiling
+    module TraceIdentifiers
+      # Used by Datadog::Profiling::TraceIdentifiers::Helper to get the trace identifiers (trace id and span id) for a
+      # given thread, if there is an active trace for that thread in Datadog.tracer.
+      class Ddtrace
+        def initialize(tracer: nil)
+          @tracer = tracer
+        end
+
+        def trace_identifiers_for(thread)
+          current_tracer = tracer
+          return unless current_tracer
+
+          correlation = current_tracer.active_correlation(thread)
+          trace_id = correlation.trace_id
+          span_id = correlation.span_id
+
+          [trace_id, span_id] if trace_id && trace_id != 0 && span_id && span_id != 0
+        end
+
+        private
+
+        def tracer
+          return @tracer if @tracer
+
+          # NOTE: Because the profiler may start working concurrently with tracer initialization,
+          # we need to be defensive here.
+          return unless Datadog.respond_to?(:tracer)
+
+          tracer = Datadog.tracer
+          return unless tracer.respond_to?(:active_correlation)
+
+          @tracer = tracer
```

Comment:
Storing the tracer can possibly hold the wrong instance of `Datadog.tracer` when `Datadog.configure` is called, as we rebuild components at that time, creating a brand new tracer instance at `Datadog.tracer`.

If this class is recreated alongside the profiler on `Datadog.configure` calls, then all is likely well *if*  `#trace_identifiers_for` cannot be invoked before the new tracer is created during `Datadog.configure`.

Reply:
Good call. This class is indeed re-created with the profiler (in helper.rb), but with the whole thing about the components lock doesn't guarantee that the new tracer will be available in `Datadog.tracer` before the profiler starts sampling. I'll tweak this a bit to avoid this issue.

Reply:
Fixed in f6711e2523d02dfcf322124ce0b51b6b6d120d3b . Let me know your thoughts, as I know this is usually not the way things are done in ddtrace ;)